blueprints/multi-interface-servers.md

```diff
@@ -0,0 +1,666 @@
+	"profileId": 10,
+	"rack": "",
+	"revalPending": false,
+	"routerHostName": "",
+	"routerPortName": "",
+	"status": "REPORTED",
+	"statusId": 3,
+	"tcpPort": 80,
+	"type": "MID",
+	"typeId": 12,
+	"updPending": false,
+	"xmppId": "mid",
+	"xmppPasswd": "",
+}
+```
+Request payloads for `PUT` and `POST` methods will need to change similarly. A
+server object should be accepted (based only on its `interfaces` property) if
+it has at least one interface, every interface has at least one IP address,
+exactly one interface has at least one IP address that is a service address, and
+no two IP addresses of the same family/version are simultaneously marked as
+'service addresses'. And, of course, the property's values must be syntactically
+and semantically valid.
+
+Below are some examples of invalid `interfaces` property arrays, to which
+Traffic Ops MUST respond with a `400 Bad Request` and an appropriate Alert:
+
+Empty array
+```json
+{"interfaces": []}
+```
+Null property
+```json
+{"interfaces": null}
+```
+Missing property
+```json
+{}
+```
+Empty `ipAddresses` sub-property
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+Null `ipAddresses` sub-property
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": null,
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+Missing `ipAddresses` sub-property
+```json
+{
+	"interfaces": [
+		{
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+No service addresses
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [
+				{
+					"address": "0.0.0.1/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": false
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+Too many interfaces with service addresses
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [
+				{
+					"address": "0.0.0.1/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": true
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		},
+		{
+			"ipAddresses": [
+				{
+					"address": "f1d0::f00d/128",
+					"gateway": "dead::beef",
+					"serviceAddress": true
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "wnlp0"
+		}
+	]
+}
+```
+Too many service addresses of the same family/version
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [
+				{
+					"address": "0.0.0.1/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": true
+				},
+				{
+					"address": "0.0.0.2/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": true
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+##### `/servers/details`
+This endpoint is subject to the same changes described above for
+`/servers`/`/servers/{{ID}}` - though obviously the full output objects will
+differ.
+
+#### Client Impact
+The structures output by various clients will change, but no client code changes
+should be necessary.
```

Comment:
ish. The actual code in `traffic_ops/client` won't need to change. The structs all live under `/lib` so they can be shared between components.

Reply:
Anyone not using one of the native clients will have to change their client code.  Maybe I'm misreading this though.

Reply:
This section is only about the client code maintained as a part of ATC. Any other clients will need to be aware of API changes across versions and adjust their code accordingly, just like always.

Reply:
+100.  Maybe this is a good push to use the client we provide and maintain as part of the project?

Reply:
Disagree, but I guess my only real objection here is wording.  We use the term client when we mean atc supported sdk.  We need to always maintain compatibility with curl so that we can better triage breakage in the API, or support simple use cases in languages not supported by ATC.

Reply:
curl support can't be broken unless we stop doing HTTP, tho

Reply:
I'm referring to the promise that the API itself follows the API versioning rules, and not just our native clients (sdk).  Additionally, acknowledgement that unsupported clients leveraging the API directly must change as this will break the API going forward.

Reply:
I'd be happy with something along the lines of adding:
```
Any unsupported 3rd-party client code will need to be updated to
leverage this new functionality in connection with these changes to core API/datamodel 
concepts going forward.
```

Reply:
But that's implicit in the versioning process? This doesn't contain a proposal to break the versioning promise. If it doesn't make it into 2.0 before release it'll need to go into 3.0 because this is a breaking API change. Clients always need to check what's changed between major releases. A blueprint is a statement of intent to do some work on the project, nothing in it should be directed at nor even directly relevant to third parties. They can just look at the changelog.